swarm/bmt/bmt.go

```diff
@@ -112,15 +112,12 @@ func NewTreePool(hasher BaseHasherFunc, segmentCount, capacity int) *TreePool {
 	// initialises the zerohashes lookup table
 	depth := calculateDepthFor(segmentCount)
 	segmentSize := hasher().Size()
-	zerohashes := make([][]byte, depth)
+	zerohashes := make([][]byte, depth+1)
 	zeros := make([]byte, segmentSize)
 	zerohashes[0] = zeros
 	h := hasher()
-	for i := 1; i < depth; i++ {
-		h.Reset()
-		h.Write(zeros)
-		h.Write(zeros)
-		zeros = h.Sum(nil)
+	for i := 1; i < depth+1; i++ {
+		zeros = doSum(h, nil, zeros, zeros)
```

Comment:
I guess this means that zero bytes don't always have the same hash? I guess that would be silly....how do we detect that they are zero bytes?

Reply:
these zerobytes array correspond to the bmt subtree hashes for 2, 4, 8, 16, ... all-zero segments